packages/lesswrong/components/ea-forum/giving-portal/ElectionCandidatesList.tsx

```diff
@@ -43,32 +60,81 @@ const sortOptions: Record<ElectionCandidatesSort, SettingsOption> = {
   },
 };
 
-const ElectionCandidatesList = ({className, classes}: {
+const sortOptions: Record<ElectionCandidatesSort | "random", SettingsOption> = {
+  ...selectSortOptions,
+  mostPreVoted: {
+    label: "Most pre-voted",
+  },
+};
+
+const ElectionCandidatesList = ({type="preVote", selectedCandidateIds, onSelect, setTotalCount, className, classes}: {
+  /**
+   * - "preVote": selecting a candidate (instantly) adds a pre-vote for it
+   * - "select": selecting a candidate runs the onSelect callback
+   */
+  type?: "preVote" | "select",
+  selectedCandidateIds?: string[],
+  onSelect?: (candidateIds: string[]) => void,
+  setTotalCount?: (count: number) => void,
   className?: string,
   classes: ClassesType,
 }) => {
-  const [sortBy, setSortBy] = useState<ElectionCandidatesSort>("mostPreVoted");
+  const isSelect = type === "select";
+  const [sortBy, setSortBy] = useState<ElectionCandidatesSort | "random">(
+    isSelect ? "random" : "mostPreVoted"
+  );
   const {results, loading} = useElectionCandidates(sortBy);
 
+  const allSelected = useMemo(
+    () => results?.every((candidate) => selectedCandidateIds?.includes(candidate._id)),
+    [results, selectedCandidateIds]
+  );
+
+  if (setTotalCount) {
+    setTotalCount(results?.length || 0);
+  }
+
   const onSelectSort = useCallback((value: string) => {
-    if (isElectionCandidateSort(value)) {
+    if (isElectionCandidateSort(value) || value === "random") {
       setSortBy(value);
     }
   }, []);
 
   const {Loading, ElectionCandidate, ForumDropdown} = Components;
   return (
     <div className={classNames(classes.root, className)}>
-      <ForumDropdown
-        value={sortBy}
-        options={sortOptions}
-        onSelect={onSelectSort}
-        className={classes.dropdown}
-      />
+      <div className={classes.controls}>
+        <ForumDropdown value={sortBy} options={isSelect ? selectSortOptions : sortOptions} onSelect={onSelectSort} className={classes.dropdown} />
+        <div className={classes.selectAll}>
+          <Checkbox
+            className={classes.checkbox}
+            style={{ color: requireCssVar("palette", "givingPortal", 1000) }}
+            checked={allSelected}
+            onChange={() => {
+              const allIds = results?.map((candidate) => candidate._id) || [];
+              const allUnselected = difference(allIds, selectedCandidateIds ?? []);
+
+              if (allUnselected.length === 0) {
+                // Clear selection
+                onSelect?.(allIds)
+              } else {
+                onSelect?.(allUnselected)
+              }
+            }}
+          />
+          Select all
+        </div>
+      </div>
       <div className={classes.grid}>
-        {loading && <Loading white />}
```

Comment:
In practice it's hard to see these loading dots in the giving portal because they are below the fold, so I didn't keep the `white` version (even if you do scroll down really quickly to see them, they look fine anyway)